R2Northstar/NorthstarLauncher
# Big rpak loading refactor

Related docs PR: https://github.com/R2Northstar/ModdingDocs/pull/251

Well this branch got out of hand.

Anyway, this PR reworks how rpaks are loaded, unloaded and tracked.

It allows for rpak reloading between map loads, meaning that skins and map overhauls could be enabled and disabled on the fly.

Previous methods of loading rpaks still work, but the new cool way is to format the `rpak.json` file like this:

```json
{
	"mp_colony_snow.rpak": "mp_colony02",
	"multiplayer_only.rpak": "mp_.*",
	"singleplayer_only.rpak": "sp_.*"
}
```

An rpak will be loaded if the map that is being loaded matches the regex given in the json.

# How to test:
## Ensuring old behaviour
1. acquire an rpak mod that uses Preload
2. acquire an rpak mod that uses Postload
3. acquire an rpak mod that uses Aliases
4. make sure that those three work as before. textures load etc.
## Testing new behaviour
1. take an rpak mod that is like a skin or something
2. edit the rpak.json to load the rpak on a map of your choice with regex
3. make sure that the rpak loads as intended
4. disable the mod or remove the entry from the rpak.json and `reload_mods`
5. go back into the same map
6. the skin should no longer be showing
### Update for testing new behaviour:
Here is an [example mod](https://github.com/user-attachments/files/16808218/Spoons.Trol.zip) that uses the new rpak loading system. It's a skin for the satchel and with the `rpak.json` setup it will only load on `mp_angel_city`

To further test this example mod, change the `mp_angel_city` to some regex that will match some maps, all maps, or whatever and test it. You shouldn't need to restart the game entirely for these changes to be respected, simply `reload_mods` and then load a level. For example `mp_` should match any multiplayer map and `.` should match every map.
